pkg/pillar/hypervisor/kvm.go

```diff
@@ -462,20 +462,135 @@ func (ctx kvmContext) Task(status *types.DomainStatus) types.Task {
+}
+
+// overhead for VMM memory mapped IO
+// it fluctuates between 0.66 and 0.81 % of MMIO total size
+// for all mapped devices. Set it to 1% to be on the safe side
+// this can be a pretty big number for GPUs with very big
+// aperture size (e.g. 64G for NVIDIA A40)
+func mmioVMMOverhead(domainName string, config types.DomainConfig, aa *types.AssignableAdapters) (int64, error) {
+	var pciAssignments []typeAndPCI
+	var mmioSize uint64 = 0
+
+	for _, adapter := range config.IoAdapterList {
+		logrus.Debugf("processing adapter %d %s\n", adapter.Type, adapter.Name)
+		list := aa.LookupIoBundleAny(adapter.Name)
+		// We reserved it in handleCreate so nobody could have stolen it
+		if len(list) == 0 {
+			return 0, logError("IoBundle disappeared %d %s for %s\n",
+				adapter.Type, adapter.Name, domainName)
+		}
+		for _, ib := range list {
+			if ib == nil {
+				continue
+			}
+			if ib.UsedByUUID != config.UUIDandVersion.UUID {
+				return 0, logError("IoBundle not ours %s: %d %s for %s\n",
+					ib.UsedByUUID, adapter.Type, adapter.Name,
+					domainName)
+			}
+			if ib.PciLong != "" {
+				logrus.Infof("Adding PCI device <%v>\n", ib.PciLong)
+				tap := typeAndPCI{pciLong: ib.PciLong, ioType: ib.Type}
+				pciAssignments = addNoDuplicatePCI(pciAssignments, tap)
+			}
+		}
+	}
+
+	for _, typ := range pciAssignments {
+		logrus.Infof("PCI device %s %d\n", typ.pciLong, typ.ioType)
+		// read the size of the PCI device aperture
+		if typ.ioType == types.IoOther {
+			// skip bridges
+			isbridge, err := typ.isBridge()
+			if err != nil {
+				// do not treat as fatal error
+				logrus.Errorf("Can't read PCI device class, treat as bridge %s: %v\n",
+					typ.pciLong, err)
+				continue
+			}
+
+			if isbridge {
+				logrus.Infof("Skipping bridge %s\n", typ.pciLong)
+				continue
+			}
+
+			// read all resources of the PCI device
+			resources, err := typ.readResources()
+			if err != nil {
+				return 0, logError("Can't read PCI device resources %s: %v\n",
+					typ.pciLong, err)
+			}
+
+			// calculate the size of the MMIO region
+			for _, res := range resources {
+				if res.valid() && res.isMem() {
+					mmioSize += res.size()
+				}
+			}
+		}
+	}
+
+	logrus.Info("MMIO size: ", mmioSize)
+	logrus.Info("MMIO overhead: ", mmioSize/100)
+
+	return int64(mmioSize / 100), nil // 1% of the total MMIO size in bytes
+}
+
+// each vCPU requires about 3MB of memory
+func cpuVMMOverhead(config types.DomainConfig) int64 {
+	return int64(config.MaxCpus) * (3 << 20) // Mb in bytes
+}
+
+// memory allocated by QEMU for its own purposes.
+// statistical analisys did not revile any correlation between
+// VM configuration (devices, nr of vcpus, etc) and this number
+// however the size of disk space affects it. Probaly some internal
+// QEMU caches are allocated based on the size of the disk image.
+// it requires more investigation.
+func undefinedVMMOverhead(config types.DomainConfig) int64 {
+	return 350 << 20 // Mb in bytes
```

Comment:
As far as I know, QEMU also takes some memory for disk caches and network buffers.
So, this number can correlate with the disk size and amount of network interfaces.

Reply:
that would be good to analyze further if you can some technical details how I can detect caches/buffers I'll really appreciate it

Reply:
Unfortunately, I don't have a direct answer here =(
Maybe you can use some commands in QEMU monitor. `info tlb`, `info mem`, `info mtree` look like something useful.